libraries/chain_kv/include/b1/chain_kv/chain_kv.hpp

```diff
@@ -110,12 +111,141 @@ inline bytes get_next_prefix(const bytes& prefix) {
    return next_prefix;
 }
 
+namespace detail {
+   typedef __uint128_t uint128_t;
+
+   template<std::size_t Size, std::size_t N>
+   void copy_and_swap_to_buffer(char* to, const char* from) {
+      char* to_ptr = to + Size * (N - 1);
+      const char* from_ptr = from;
+      for (unsigned int i = 0; i < N; ++i) {
+         memcpy(to_ptr, from_ptr, Size);
+         to_ptr -= Size;
+         from_ptr += Size;
+      }
+   }
+
+   template<typename T>
+   struct value_storage;
+
+   template<typename T>
+   struct value_storage<T*> {
+      char* as_char_ptr(T* value) const { return reinterpret_cast<char*>(value); }
+      constexpr std::size_t size() const { return sizeof(T); }
+   };
+
+   template<typename T>
+   struct value_storage {
+      char* as_char_ptr(T& value) const { return reinterpret_cast<char*>(&value); }
+      constexpr std::size_t size() const { return sizeof(T); }
+   };
+
+   template <typename T, std::size_t N>
+   auto append_key(bytes& dest, T value) {
+      constexpr auto type_size = value_storage<T>().size();
+      constexpr auto buf_size = type_size * N;
+      char buf[buf_size];
+      detail::copy_and_swap_to_buffer<type_size, N>(buf, value_storage<T>().as_char_ptr(value));
+      std::reverse(std::begin(buf), std::end(buf));
+      dest.insert(dest.end(), std::begin(buf), std::end(buf));
+   }
+
+   template <typename UInt, typename T>
+   UInt float_to_key(T value) {
+      static_assert(sizeof(T) == sizeof(UInt), "Expected unsigned int of the same size");
+      UInt result;
+      std::memcpy(&result, &value, sizeof(T));
+      const UInt signbit = (static_cast<UInt>(1) << (std::numeric_limits<UInt>::digits - 1));
+      UInt mask    = 0;
+      if (result == signbit)
+         result = 0;
+      if (result & signbit)
+         mask = ~mask;
+      return result ^ (mask | signbit);
+   }
+
+   template <typename T, typename UInt>
+   T key_to_float(UInt value) {
+      static_assert(sizeof(T) == sizeof(UInt), "Expected unsigned int of the same size");
+      // encoded signbit indicates positive value
+      const UInt signbit = (static_cast<UInt>(1) << (std::numeric_limits<UInt>::digits - 1));
+      UInt mask    = 0;
+      if ((value & signbit) == 0)
+         mask = ~mask;
+      value = {value ^ (mask | signbit)};
+      T float_result;
+      std::memcpy(&float_result, &value, sizeof(UInt));
+      return float_result;
+   }
+
+   template<typename Key, std::size_t N>
+   bool extract_key(bytes::const_iterator& key_loc, bytes::const_iterator key_end, Key& key) {
+      const auto distance = std::distance(key_loc, key_end);
+      constexpr auto type_size = value_storage<Key>().size();
+      constexpr auto key_size = type_size * N;
+      if (distance < key_size)
+         return false;
+
+      key_end = key_loc + key_size;
+      bytes key_store(key_loc, key_end);
```

Comment:
It seems like there might be a solid use case for a reverse iterator on bytes.  Might save use from needing to do the `std::reverse` in some places.